test/integration/client-side-encryption/client_side_encryption.prose.test.js

```diff
@@ -2046,4 +2046,124 @@ describe('Client Side Encryption Prose Tests', metadata, function () {
+    const masterKeys = {
+      aws: {
+        region: 'us-east-1',
+        key: 'arn:aws:kms:us-east-1:579766882180:key/89fcc2c4-08b0-4bd9-9f25-e30687b580d0'
+      },
+      azure: {
+        keyVaultEndpoint: 'key-vault-csfle.vault.azure.net',
+        keyName: 'key-name-csfle'
+      },
+      gcp: {
+        projectId: 'devprod-drivers',
+        location: 'global',
+        keyRing: 'key-ring-csfle',
+        keyName: 'key-name-csfle'
+      },
+      kmip: {},
+      local: undefined
+    };
+    let client1, client2;
+
+    /**
+     * Run the following test case for each pair of KMS providers (referred to as ``srcProvider`` and ``dstProvider``).
+     * Include pairs where ``srcProvider`` equals ``dstProvider``.
+     */
+    function* generateTestCombinations() {
+      const providers = Object.keys(masterKeys);
+      for (const srcProvider of providers) {
+        for (const dstProvider of providers) {
+          yield { srcProvider, dstProvider };
+        }
+      }
+    }
+
+    beforeEach(async function () {
+      client1 = await this.configuration.newClient().connect();
+      client2 = await this.configuration.newClient().connect();
+    });
+
+    afterEach(async function () {
+      await client1.close();
+      await client2.close();
+    });
+
+    for (const { srcProvider, dstProvider } of generateTestCombinations()) {
+      it(
+        `should rewrap data key from ${srcProvider} to ${dstProvider}`,
+        metadata,
+        async function () {
+          // Step 1. Drop the collection ``keyvault.datakeys``
+          await client1
+            .db('keyvault')
+            .dropCollection('datakeys')
+            .catch(() => null);
+
+          // Step 2. Create a ``ClientEncryption`` object named ``clientEncryption1``
+          const clientEncryption1 = new this.configuration.mongodbClientEncryption.ClientEncryption(
+            client1,
+            {
+              keyVaultNamespace: 'keyvault.datakeys',
+              kmsProviders: getKmsProviders(),
+              tlsOptions: {
+                kmip: {
+                  tlsCAFile: process.env.KMIP_TLS_CA_FILE,
+                  tlsCertificateKeyFile: process.env.KMIP_TLS_CERT_FILE
+                }
+              }
+            }
+          );
+
+          // Step 3. Call ``clientEncryption1.createDataKey`` with ``srcProvider``
+          const keyId = await clientEncryption1.createDataKey(srcProvider, {
+            masterKey: masterKeys[srcProvider]
+          });
+
+          // Step 4. Call ``clientEncryption1.encrypt`` with the value "test"
+          const cipherText = await clientEncryption1.encrypt('test', {
+            keyId,
+            algorithm: 'AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic'
+          });
+
+          // Step 5. Create a ``ClientEncryption`` object named ``clientEncryption2``
+          const clientEncryption2 = new this.configuration.mongodbClientEncryption.ClientEncryption(
+            client1,
```

Comment:
I think this should technically be

```suggestion
            client2,
```

That being said … I’m not sure how much this really matters. In terms of adhering to the spec test, strictly speaking the test would need to create a new pair of clients for *every combination* of providers. If that’s something we don’t want to do, because it’s slow or really for whatever reason, then that reason probably can be extended to just using a single client for all ClientEncryption objects as well.

Reply:
Good catch! 

We actually do already create two clients for each pair of providers - it's done in the beforeEach hook for the test.

Reply:
Generally in our tests we create a new client for each test.  I don't know if that slows down our tests substantially but I'd guess that it doesn't matter much.  My guess is that since our tests use a real server, the bottleneck is really network i/o with the server and the server writing to the disk, not our actual test code.  It's worth thinking about though if our tests start taking a long time to run.